dask/dataframe/groupby.py

```diff
@@ -132,6 +136,363 @@ def _nunique_series_aggregate(df):
     return df.groupby(df.columns[1])[df.columns[0]].nunique()
 
 
+###############################################################
+# Aggregate support
+#
+# Aggregate is implemented as:
+#
+# 1. group-by-aggregate all partitions into intermediate values
+# 2. collect all partitions into a single partition
+# 3. group-by-aggregate the result into intermediate values
+# 4. transform all intermediate values into the result
+#
+# In Step 1 and 3 the dataframe is grouped on the same columns.
+#
+###############################################################
+
+def _make_agg_id_factory():
+    """
+    Return a factory for ids in the aggregate graph.
+
+    Keys are built from the function, the column to apply the function to, and
+    a increasing numeric id. The returns ids are deterministic. However, the
+    order in which ids are generated changes the numeric id. For different
+    calls with the same arguments, the factory function will always return the
+    same id.
+    """
+    integers = it.count()
+    ids = {}
+
+    def factory(func, column):
```

Comment:
Why not just use `dask.base.tokenize`? This provides deterministic keys, without the need for cache.

```
tokenize(func, column)
```